Immutable arrays

goldfirere opened the pull request:
I've created this PR instead of working in https://github.com/ocaml-flambda/ocaml-jst/pull/47 so that the comprehensions stuff doesn't get in the way.

antalsz committed:
Immutable arrays

goldfirere commented on stdlib/array.ml:
```diff
@@ -162,6 +162,15 @@ let of_list = function
         | hd::tl -> unsafe_set a i hd; fill (i+1) tl in
       fill 1 tl
 
+(* Immutable and mutable arrays have the same runtime representation; we
```
@mshinwell, could you confirm that this is OK? I see that mutability is tracked in Lambda, but indeed it seems not to change the translation away from Lambda, so this is likely OK. But `Obj.magic` requires extra scrutiny. Thanks!

mshinwell replied:
I would just write out the copying function, it's small, then we don't need `Obj.magic`.

goldfirere commented on stdlib/array.mli:
```diff
@@ -24,6 +24,13 @@ open! Stdlib
    arrayLabels.mli instead.
  *)
 
+(* CR aspectorzabusky: When I ran [tools/sync_stdlib_docs], it copied the
```
@stedolan, I'm sure you know the answer to this...

goldfirere commented on stdlib/array.ml:
```diff
@@ -162,6 +162,15 @@ let of_list = function
+   1. doesn't mutate its input;
+   2. doesn't hold on to its input; and
+   3. returns a fresh array as its output. *)
+let to_iarray = Obj.magic (copy : 'a array -> 'a array)
```
Maybe I'm just paranoid, but I'd prefer a `let to_iarray : 'a array -> 'a iarray = ...` here. Yes, it's redundant with the mli file. But it would be Bad to change the mli file's type signature and then quietly make this unsound.

antalsz replied:
This would have been nice, but ran into the value restriction.  I've added a comment instead.

goldfirere commented on stdlib/array.ml:
```diff
@@ -162,6 +162,15 @@ let of_list = function
+   2. doesn't hold on to its input; and
+   3. returns a fresh array as its output. *)
+let to_iarray = Obj.magic (copy : 'a array -> 'a array)
+let of_iarray = Obj.magic (copy : 'a array -> 'a array)
```
These functions make me want

```ocaml
type immutable = |
type mutable = |
type ('a, 'mut) gen_array

type 'a array = ('a, mutable) gen_array
type 'a iarray = ('a, immutable) gen_array
```

and then many functions (including user-written ones) can be polymorphic over mutability. This also means that we could drop the `Obj.magic` here.

Actually, this makes me think: immutability could be done entirely in a library. The only reason to involve the compiler at all is 1) to perform optimizations and 2) to provide good syntax. Would this approach simplify things?

goldfirere replied:
I suppose this would destroy the nice covariance annotation. But is that worth it? Or maybe I'm missing something more fundamental.

goldfirere replied:
After more thought, this is a move in the wrong direction: covariance is nice indeed, and there is the possibility for efficient array slicing for immutable arrays.

goldfirere commented on stdlib/iarray.ml:
```diff
@@ -0,0 +1,51 @@
+(* CR aspectorzabusky: This needs a copyright header; should I copy [array.ml]? *)
```
@lpw25 any thoughts?

mshinwell replied:
Are these files completely new?  If so they should have JS copyright headers, you can crib from here: https://github.com/janestreet/base/blob/master/LICENSE.md

goldfirere commented on stdlib/iarray.ml:
```diff
@@ -0,0 +1,51 @@
+
+[@@@ocaml.flambda_o3]
+
+(* An alias for the type of arrays. *)
```
```suggestion
(* An alias for the type of immutable arrays. *)
```

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+   If the return type of [f] is [float], then the maximum
+   size is only [Sys.max_array_length / 2]. *)
+
+val make_matrix :
```
This function seems pretty unuseful.

antalsz replied:
Done, and I added a comment about it.

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+   designate a valid subarray of [a]; that is, if
+   [pos < 0], or [len < 0], or [pos + len > length a]. *)
+
+(* CR aspectorzabusky: I dropped [copy] because these are immutable.  Is there
```
No, I don't think so. Leave it out.

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+val concat : 'a iarray list -> 'a iarray
+(** Same as {!append}, but concatenates a list of immutable arrays. *)
+
+val sub : 'a iarray -> pos:int -> len:int -> 'a iarray
```
Does this copy? It does not need to. Actually, it's a bit of a shame that it does copy. The ability to slice an immutable array is one of its advantages over a mutable one. In any case, if it copies, the documentation should say so.

antalsz replied:
As discussed, it does have to copy; I've added a sentence saying so.

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+(** [create_float n] returns a fresh immutable float array of length [n],
+    with uninitialized data. *)
+
+val init : int -> f:(int -> 'a) -> 'a iarray
```
The `f` argument can be marked as local. Actually, a lot of arguments here can be local. Perhaps it would be good to imagine how to make this interface more local-friendly. Could also be a good topic for a type-systems meeting.

antalsz replied:
We decided not to make these functions local in the stdlib and only do that in Base

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+
+(** Operations on immutable arrays. *)
+
+type +'a t = 'a iarray
```
Is it possible to put an array on the stack? If so, it would be good to annotate locality information on the functions here.

antalsz replied:
We decided not to do this in the stdlib and only do this in Base

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+
+(** {1 Converting to and from mutable arrays} *)
+
+(* CR aspectorzabusky: When we add locals, we can do
```
We have added locals!

antalsz replied:
It turns out we haven't made enough of the `Array` API `local_` for these functions to be useful yet; deferred to the future.

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+external ( .:() ) : 'a iarray -> int -> 'a = "%array_safe_get"
+(** A synonym for [get]. *)
+
+external make : int -> 'a -> 'a iarray = "caml_make_vect"
```
This seems not very useful.

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+   If the value of [x] is a floating-point number, then the maximum
+   size is only [Sys.max_array_length / 2]. *)
+
+external create_float: int -> float iarray = "caml_make_float_vect"
```
This seems positively unuseful.

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+   function is applied to the index of the element as first argument,
+   and the element itself as second argument. *)
+
+val fold_left : f:('a -> 'b -> 'a) -> init:'a -> 'b iarray -> 'a
```
4.14 also comes with `fold_left_map`, which should be added here, too.

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+val memq : 'a -> set:'a iarray -> bool
+(** Same as {!mem}, but uses physical equality
+   instead of structural equality to compare list elements. *)
+
```
Also add 4.14's `find_opt`, `find_map`, `split`, and `combine`.

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+(** Same as {!mem}, but uses physical equality
+   instead of structural equality to compare list elements. *)
+
+(* CR aspectorzabusky: We should add non–in-place sorting *)
```
Yes.

goldfirere commented on stdlib/iarrayLabels.mli:
```diff
@@ -0,0 +1,221 @@
+   instead of structural equality to compare list elements. *)
+
+(* CR aspectorzabusky: We should add non–in-place sorting *)
+
```
Might be nice to have some kind of

```ocaml
(** Create an immutable array by filling in a mutable array. This is the same as building a mutable array
and then using [of_array], but without a copy. *)
val make_mutably : int -> (local_ 'a array -> unit) -> 'a iarray
```

This would then allow users to e.g. write `sort`. Note that without the `local_`, this would be wildly unsafe! But I think it's safe with the `local_`, given that all mutable fields must be global. (The closure itself could also be labeled as `local_`, but that's a separate matter.)

antalsz replied:
These were the functions discussed above.

goldfirere commented on typing/outcometree.mli:
```diff
@@ -40,6 +40,7 @@ type out_attribute =
 
 type out_value =
   | Oval_array of out_value list
+  | Oval_iarray of out_value list (* CR aspectorzabusky: Or [Oval_immutable_array]? *)
```
`iarray` is fine. But actually, I think it would be slightly better to have `Oval_array of out_value list * Asttypes.mutable_flag`.

goldfirere commented on typing/parmatch.ml:
```diff
@@ -302,7 +302,9 @@ module Compat
-  | Tpat_array ps, Tpat_array qs ->
+  | Tpat_array (am1, ps), Tpat_array (am2, qs) ->
+      am1 = am2 &&
+      (* CR aspectorzabusky: Can't we drop the length check here? *)
```
I don't think so. Two patterns are compatible if there exists a value that matches both, as per "Compatibility check" above.

antalsz replied:
It would be correct to delete the length check but it's an optimization, because arrays have O(1) length

goldfirere commented on typing/typecore.ml:
```diff
@@ -1683,6 +1691,32 @@ and type_pat_aux
+      (* Normally this would go to an auxiliary function, but this function
+         takes so many parameters, has such a complex type, and uses so many
+         local definitions, it seems better to just put the pattern matching
+         here.  This shouldn't fuck up the diff *too* much. *)
```
Maybe I'm just a bit curmudgeonly, but I'd prefer cleaner language in the codebase.

goldfirere commented on typing/typecore.ml:
```diff
@@ -2890,6 +2918,13 @@ let is_local_returning_expr e =
         raise(Error(loc2, Env.empty, Local_return_annotation_mismatch loc1))
   in
   let rec loop e =
+    match Extensions.Expression.of_ast e with
```
Seeing this inspired me to look for other `Parsetree.expression` matches. I found one in `parsing/ast_invariants.ml` (line 80) that should probably get a call to `of_ast`.

antalsz replied:
This actually raises an interesting can of worms: what do we do with `ast_iterator`?  I think we have to leave it alone, right?

goldfirere replied:
"Have to"? No. We could add a little hook that processes immutable arrays in some way. But there is no good reason to do so -- note that arrays simply process their subexpressions. So we should do the same for immutable arrays, which happens automatically.

antalsz replied:
Also, fixing this triggered a test failure; please take a look at the solution in `testsuite/tests/ast-invariants/test.ml`, I'm not sure if it's the best approach

goldfirere commented on typing/typecore.ml:
```diff
@@ -5896,7 +5931,7 @@ and type_let
     | _ -> false
-  and eexp_is_fun : Extensions.extension_expr -> _ = function
+  and eexp_is_fun : Extensions.Expression.t -> _ = function
     | _ -> false
```
Maybe avoid the `_ ->` here?

goldfirere commented on typing/typecore.ml:
```diff
@@ -6510,6 +6585,20 @@ and type_comprehension_iterator
+        ~explanation
+        ~type_:Predef.type_iarray
+        ~mutability:Immutable
+        ~attributes:[] (* CR aspectorzabusky: This can't be right *)
```
You're right -- this is problematic. The problem is in the parsing scheme, which doesn't really put the attributes anywhere.

antalsz replied:
This will be dealt with in the extensions PR

goldfirere commented on typing/typecore.ml:
```diff
@@ -6656,8 +6745,11 @@ let report_type_expected_explanation expl ppf =
-         | List_comprehension  -> "a list"
-         | Array_comprehension -> "an array")
+         | List_comprehension   -> "a list"
+         | Array_comprehension  -> "an array" (* CR aspectorzabusky: We should
```
Yes, skip "mutable".

goldfirere commented on typing/typecore.mli:
```diff
@@ -23,6 +23,7 @@ open Types
 type comprehension_type =
   | List_comprehension
   | Array_comprehension
```
It's not clear to me whether this is better or `Array_comprehension of Asttypes.mutable_flag` is better. Just throwing out another possibility -- it's perfectly fine to keep this as-is, too.

antalsz replied:
I think moving pretty uniformly to the `mutable_flag` form is better; change made

goldfirere commented on typing/typedtree.mli:
```diff
@@ -113,8 +113,13 @@ and 'k pattern_desc =
+      mutable_flag * value general_pattern list -> value pattern_desc
+        (** [| P1; ...; Pn |]    (flag = Mutable)
+            [: P1; ...; Pn :]    (flag = Immutable) *)
+    (* CR aspectorzabusky: This ended up feeling clearer to me than two
```
I like this design

antalsz replied:
Yes, moving this way everywhere.

goldfirere commented on typing/typedtree.mli:
```diff
@@ -248,9 +253,13 @@ and expression_desc =
       expression * Longident.t loc * Types.label_description * expression
-  | Texp_array of expression list
+  | Texp_array of mutable_flag * expression list
+    (* CR aspectorzabusky: This ended up feeling clearer to me than two
```
I like this design. If anything, I would want more of this kind of design for this feature, though you have the experience to make the better decision.

antalsz replied:
I think you're correct here.

goldfirere commented on lambda/transl_array_comprehension.mli:
```diff
@@ -4,12 +4,17 @@ open Debuginfo.Scoped_location
 
 (** Translate array comprehensions; see the .ml file for more details *)
 
+(* CR aspectorzabusky: Is it really true that I don't need to pass in array
```
@mshinwell? This code, as written, will result in `Pmakearray(..., Mutable, ...)`, even for an immutable array. This makes sense, because the comprehensions code actually mutates the memory after creation. But then we'll have a mutable block as the value for an immutable array. I don't quite see what goes wrong here, but I'd love more scrutiny on this point. (@antalsz: Given that Mark is away for the next few weeks, feel free to find someone else who can answer this question.)

mshinwell replied:
I think this is correct.  All that should happen is that the middle-end won't be able to track anything about the contents of the array.

goldfirere commented on lambda/translcore.ml:
```diff
@@ -31,7 +31,7 @@ type error =
 exception Error of Location.t * error
 
-let use_dup_for_constant_arrays_bigger_than = 4
+let use_dup_for_constant_mutable_arrays_bigger_than = 4
```
I'm scared of this module, because
1. I don't know the ramifications of the `Mutable`/`Immutable` flag in lambda.
2. Getting this wrong seems like it could be catastrophic.
@mshinwell is an obvious reviewer for this module, but others may also do well.

mshinwell replied:
I'll read the changes to this module (hopefully tomorrow).
We should try to arrange things so that immutable arrays at the source level (except for ones constructed with comprehensions) turn into immutable arrays in Lambda -- this means that e.g. projection from array literals can be evaluated by the middle-end.  I'm unsure if this is happening at present.

mshinwell replied:
(and also it means that things like nested immutable array literals can be statically allocated)

mshinwell replied:
@goldfirere @antalsz I've read these `Translcore` changes, I think they're ok - but we should test carefully, this stuff is tricky.

goldfirere commented on parsing/extensions.ml:
```diff
@@ -1,265 +1,46 @@
-    directions for all new syntax, lowering it to extension nodes and then
-    (somewhat more obnoxiously) lifting it back out. *)
-
+open Asttypes
```
I am not reviewing the extensions architecture. See #1.

goldfirere commented on parsing/parser.mly:
```diff
@@ -2351,6 +2378,7 @@ simple_expr:
       { array_get ~loc:$sloc $1 $4 }
   | simple_expr DOT LPAREN seq_expr error
       { unclosed "(" $loc($3) ")" $loc($5) }
+  (* Immutable array indexing is a regular operator *)
```
I don't understand the import of this comment.

antalsz replied:
This should've gotten clarified in the rebase, there are a couple new comments about immutable array indexing now

goldfirere commented on parsing/parser.mly:
```diff
@@ -2496,20 +2542,20 @@ comprehension_clause:
-    LBRACKETBAR expr_semi_list error
-      { unclosed "[|" $loc($3) "|]" $loc($5) }
+  | array_exprs(LBRACKETBAR, BARRBRACKET)
+      { Generic_array.expression
```
Is `Generic_array` really necessary here? Can you not pass ml values into a parameterized non-terminal? What if you had a function that branched on the open/close terminal? If `Generic_array` stays, though, then it needs more documentation.

goldfirere reviewed (commented):
Looks pretty good! There are a few loose ends, as you commented, and a few parts I was not competent to review.

antalsz committed:
Make `comprehensions` into `comprehensions_experimental` for now (#109)

antalsz committed:
Temporarily disable list comprehensions tests due to locals bug

goldfirere committed:
Fix test case to use comprehensions_experimental

goldfirere committed:
Skip failing test for now

goldfirere committed:
Consistently use Lmutvar or Lvar in comprehensions (#111)

This commit adds an [occur] field to Let_binding so that
it can correctly use Lvar or Lmutvar. Tiny bonus: fewer
allocations from Lvar nodes, because now we just create one.

goldfirere committed:
Better error message for under-applied functions (#74)

antalsz committed:
Make `immutable_arrays` into `immutable_arrays_experimental` for now

riaqn committed:
fix several issues when removing exp_mode (#110)

* fix several issues when removing exp_mode

goldfirere committed:
Add stdlib iarray files to dune file

mshinwell commented on stdlib/iarray.ml:
```diff
@@ -0,0 +1,95 @@
+   1. None of these functions mutate their array inputs;
+   2. None of these functions hold on to their array inputs; and
+   3. All of these functions return fresh arrays if they return an array. *)
+let init =
```
Personally I'd rather avoid uses of `Obj.magic` here -- there are very few uses in the stdlib, which seems like a good thing.  Can we not just copy the code of the original `Array` module and edit it as required (in which case the INRIA copyright notice must be preserved)?

mshinwell commented on lambda/lambda.mli:
```diff
@@ -38,15 +38,23 @@ type alloc_mode = private
   | Alloc_heap
   | Alloc_local
 
+type modify_mode = private
```
I'm unsure why the change to add this `modify_mode` is in this PR.  Could we split this into a separate PR?

mshinwell commented on lambda/translcore.ml:
```diff
@@ -509,47 +503,58 @@ and transl_exp0 ~in_new_scope ~scopes e =
+                imm_array
               else
+                (* CR aspectorzabusky: Do we construct things correctly in this
+                   case? *)
```
I think this is ok.

mshinwell commented on lambda/translcore.ml:
```diff
@@ -558,18 +563,21 @@ and transl_exp0 ~in_new_scope ~scopes e =
         ~transl_exp ~scopes ~loc comp
-  | Texp_array_comprehension comp ->
+  | Texp_array_comprehension (_amut, comp) ->
+      (* We can ignore mutability here since we've already checked in in the
```
Typo in comment

antalsz commented:
This got resolved on ocaml-jst
